File: leetcode/ag_123.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        ln = len(prices)
        if ln <= 1: return 0
        dp = []
        k = 2
        for y in xrange(k+1):
            dp.append([])
            for x in xrange(ln): 
                dp[y].append(0)
        for i in range(1,k+1):
            maxv = dp[i-1][0] - prices[0]
            for t in range(1, ln):
                dp[i][t] = max(dp[i][t-1], maxv + prices[t])
                maxv = max(maxv, dp[i-1][t] - prices[t])
        return dp[k][ln-1]


def run():
    s = Solution()
    t = [3,3,5,0,0,3,1,4]
    p = s.maxProfit(t)
    print p
        

File: leetcode/ag_123.cpp
#include <iostream>
#include <vector>
using namespace std;



class Solution { 
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size(), num = 2;
        if (n <= 1) return 0;
        vector<vector<int> > dp(num + 1, vector<int>(n, 0));
        for (int k = 1; k <= num; k++) {
            int temp = dp[k - 1][0] - prices[0];
            for (int i = 1; i < n; i++) {
                dp[k][i] = max(dp[k][i - 1], prices[i] + temp);
                temp = max(temp, dp[k - 1][i] - prices[i]);
            }
        }
        for(int i = 0; i < num+1; i++) {
            for (int j = 0; j < n; j++) {
                cout<<dp[i][j]<<" ";
            }
            cout<<endl;
        }
        return dp[num][n - 1];
    } 
} ss;
int main()
{
    int input[] = {3,3,5,0,0,3,1,4 };
    vector<int> s(input, input + sizeof(input)/sizeof(int));
    int ans = ss.maxProfit(s);
    cout<<"ans is"<<ans<<endl;
    return 0;
}
